File: supabase/functions/_shared/blueprint-importer/estimate-mapping.ts
// Blueprint Importer v2 — Estimate mapping contracts (Phase 1).
// Side-effect-free type module. No DB, no IO, no estimate generation.

import type { TradeId } from "./trade-catalog.ts";
import type { BlueprintMeasurementObject } from "./measurement-objects.ts";
import type { BlueprintPlanPath } from "./plan-path.ts";

export type ImportSessionStatus =
  | "draft"
  | "parsed"
  | "trades_detected"
  | "user_review_required"
  | "accepted"
  | "rejected"
  | "superseded"
  | "failed";

export type SourceContextType =
  | "project"
  | "opportunity"
  | "lead"
  | "estimate"
  | "contact"
  | "standalone";

export type DocumentType =
  | "roof_report"
  | "wall_report"
  | "blueprint_set"
  | "spec_book"
  | "addendum"
  | "unknown";

export type SourceProvider =
  | "roofr"
  | "eagleview"
  | "internal_geometry"
  | "user_uploaded_blueprint"
  | "unknown";

export interface BlueprintImportSession {
  id?: string;
  tenant_id: string;
  source_context_type: SourceContextType;
  source_context_id?: string | null;
  status: ImportSessionStatus;
  contract_version: string;
  deterministic_hash?: string | null;
  notes?: string | null;
  metadata?: Record<string, unknown>;
  created_by?: string | null;
}

export interface BlueprintSourceDocument {
  id?: string;
  import_session_id: string;
  tenant_id: string;
  file_id?: string | null;
  storage_path?: string | null;
  document_reference?: string | null;
  document_type: DocumentType;
  provider: SourceProvider;
  original_filename?: string | null;
  page_count?: number | null;
  report_date?: string | null;
  property_address?: string | null;
  property_latitude?: number | null;
  property_longitude?: number | null;
  content_hash?: string | null;
  extraction_status: "pending" | "in_progress" | "succeeded" | "failed" | "skipped";
  metadata?: Record<string, unknown>;
}

export interface BlueprintDetectedTrade {
  id?: string;
  import_session_id: string;
  tenant_id: string;
  trade_id: TradeId;
  support_status: "mvp_supported" | "measurement_object_only" | "future_supported" | "unsupported";
  confidence: number; // 0..1
  detection_signals?: Record<string, unknown>;
  source_document_ids?: string[];
  status: "detected" | "dismissed" | "superseded" | "promoted";
}

export interface BlueprintAcceptedTrade {
  id?: string;
  import_session_id: string;
  tenant_id: string;
  detected_trade_id?: string | null;
  trade_id: TradeId;
  accepted_by?: string | null;
  accepted_at?: string;
  status: "accepted" | "rejected" | "superseded";
  selected_template_id?: string | null;
  user_assumptions?: Record<string, unknown>;
  review_state: "pending_review" | "blocked" | "cleared" | "manual_only";
}

export interface BlueprintTemplateBinding {
  id?: string;
  import_session_id: string;
  tenant_id: string;
  accepted_trade_id: string;
  trade_id: TradeId;
  template_id?: string | null;
  template_version?: string | null;
  binding_status: "pending" | "ready" | "blocked" | "rejected" | "superseded";
  required_inputs: Record<string, unknown>;
  optional_inputs?: Record<string, unknown>;
  missing_inputs?: string[];
  user_assumptions?: Record<string, unknown>;
}

export interface BlueprintMaterialDraftLine {
  id?: string;
  import_session_id: string;
  tenant_id: string;
  accepted_trade_id: string;
  template_binding_id?: string | null;
  material_rule_id?: string | null;
  item_key: string;
  item_name?: string | null;
  quantity?: number | null;
  unit?: string | null;
  rounding_rule?: string | null;
  waste_percent?: number | null;
  source_measurement_ids: string[];
  plan_path_ids: string[];
  formula_key?: string | null;
  formula_inputs?: Record<string, unknown>;
  catalog_resolution_status: "unresolved" | "matched" | "ambiguous" | "missing" | "manual_override";
  catalog_item_id?: string | null;
  status: "draft" | "ready" | "blocked" | "rejected" | "superseded";
}

export interface BlueprintLaborDraftLine {
  id?: string;
  import_session_id: string;
  tenant_id: string;
  accepted_trade_id: string;
  template_binding_id?: string | null;
  labor_rule_id?: string | null;
  labor_key: string;
  labor_name?: string | null;
  quantity?: number | null;
  unit?: string | null;
  base_rate?: number | null;
  complexity_multiplier?: number | null;
  source_measurement_ids: string[];
  plan_path_ids: string[];
  formula_key?: string | null;
  formula_inputs?: Record<string, unknown>;
  status: "draft" | "ready" | "blocked" | "rejected" | "superseded";
}

// Re-export the measurement object + plan path types so downstream consumers
// only need to import from the package barrel.
export type { BlueprintMeasurementObject, BlueprintPlanPath };

File: worker/app/blueprint_contracts/estimate_mapping.py
"""Estimate mapping contract twins (Phase 1). Side-effect free."""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Literal, Optional

ImportSessionStatus = Literal[
    "draft",
    "parsed",
    "trades_detected",
    "user_review_required",
    "accepted",
    "rejected",
    "superseded",
    "failed",
]

SourceContextType = Literal[
    "project", "opportunity", "lead", "estimate", "contact", "standalone"
]

DocumentType = Literal[
    "roof_report", "wall_report", "blueprint_set", "spec_book", "addendum", "unknown"
]

SourceProvider = Literal[
    "roofr", "eagleview", "internal_geometry", "user_uploaded_blueprint", "unknown"
]


@dataclass
class BlueprintImportSession:
    tenant_id: str
    source_context_type: SourceContextType
    status: ImportSessionStatus = "draft"
    contract_version: str = "blueprint-importer-v2"
    source_context_id: Optional[str] = None
    deterministic_hash: Optional[str] = None
    notes: Optional[str] = None
    metadata: dict[str, Any] = field(default_factory=dict)
    created_by: Optional[str] = None
    id: Optional[str] = None


@dataclass
class BlueprintSourceDocument:
    import_session_id: str
    tenant_id: str
    document_type: DocumentType
    provider: SourceProvider
    extraction_status: Literal[
        "pending", "in_progress", "succeeded", "failed", "skipped"
    ] = "pending"
    file_id: Optional[str] = None
    storage_path: Optional[str] = None
    document_reference: Optional[str] = None
    original_filename: Optional[str] = None
    page_count: Optional[int] = None
    report_date: Optional[str] = None
    property_address: Optional[str] = None
    property_latitude: Optional[float] = None
    property_longitude: Optional[float] = None
    content_hash: Optional[str] = None
    metadata: dict[str, Any] = field(default_factory=dict)
    id: Optional[str] = None


@dataclass
class BlueprintDetectedTrade:
    import_session_id: str
    tenant_id: str
    trade_id: str
    support_status: str
    confidence: float = 0.0
    detection_signals: dict[str, Any] = field(default_factory=dict)
    source_document_ids: list[str] = field(default_factory=list)
    status: Literal["detected", "dismissed", "superseded", "promoted"] = "detected"
    id: Optional[str] = None


@dataclass
class BlueprintAcceptedTrade:
    import_session_id: str
    tenant_id: str
    trade_id: str
    review_state: Literal["pending_review", "blocked", "cleared", "manual_only"] = "pending_review"
    status: Literal["accepted", "rejected", "superseded"] = "accepted"
    detected_trade_id: Optional[str] = None
    accepted_by: Optional[str] = None
    accepted_at: Optional[str] = None
    selected_template_id: Optional[str] = None
    user_assumptions: dict[str, Any] = field(default_factory=dict)
    id: Optional[str] = None


@dataclass
class BlueprintTemplateBinding:
    import_session_id: str
    tenant_id: str
    accepted_trade_id: str
    trade_id: str
    binding_status: Literal["pending", "ready", "blocked", "rejected", "superseded"] = "pending"
    required_inputs: dict[str, Any] = field(default_factory=dict)
    optional_inputs: dict[str, Any] = field(default_factory=dict)
    missing_inputs: list[str] = field(default_factory=list)
    user_assumptions: dict[str, Any] = field(default_factory=dict)
    template_id: Optional[str] = None
    template_version: Optional[str] = None
    id: Optional[str] = None


@dataclass
class BlueprintMaterialDraftLine:
    """Schema-only twin; NOT populated in Phase 1."""
    import_session_id: str
    tenant_id: str
    accepted_trade_id: str
    item_key: str
    source_measurement_ids: list[str] = field(default_factory=list)
    plan_path_ids: list[str] = field(default_factory=list)
    template_binding_id: Optional[str] = None
    material_rule_id: Optional[str] = None
    item_name: Optional[str] = None
    quantity: Optional[float] = None
    unit: Optional[str] = None
    rounding_rule: Optional[str] = None
    waste_percent: Optional[float] = None
    formula_key: Optional[str] = None
    formula_inputs: dict[str, Any] = field(default_factory=dict)
    catalog_resolution_status: Literal[
        "unresolved", "matched", "ambiguous", "missing", "manual_override"
    ] = "unresolved"
    catalog_item_id: Optional[str] = None
    status: Literal["draft", "ready", "blocked", "rejected", "superseded"] = "draft"
    id: Optional[str] = None


@dataclass
class BlueprintLaborDraftLine:
    """Schema-only twin; NOT populated in Phase 1."""
    import_session_id: str
    tenant_id: str
    accepted_trade_id: str
    labor_key: str
    source_measurement_ids: list[str] = field(default_factory=list)
    plan_path_ids: list[str] = field(default_factory=list)
    template_binding_id: Optional[str] = None
    labor_rule_id: Optional[str] = None
    labor_name: Optional[str] = None
    quantity: Optional[float] = None
    unit: Optional[str] = None
    base_rate: Optional[float] = None
    complexity_multiplier: Optional[float] = None
    formula_key: Optional[str] = None
    formula_inputs: dict[str, Any] = field(default_factory=dict)
    status: Literal["draft", "ready", "blocked", "rejected", "superseded"] = "draft"
    id: Optional[str] = None
